pkg/datapath/sockets/sockets.go

```diff
@@ -0,0 +1,294 @@
+	States   uint32
+	ID       netlink.SocketID
+}
+
+type writeBuffer struct {
+	Bytes []byte
+	pos   int
+}
+
+func (b *writeBuffer) write(c byte) {
+	b.Bytes[b.pos] = c
+	b.pos++
+}
+
+func (b *writeBuffer) next(n int) []byte {
+	s := b.Bytes[b.pos : b.pos+n]
+	b.pos += n
+	return s
+}
+
+func (r *socketRequest) Serialize() []byte {
+	b := writeBuffer{Bytes: make([]byte, sizeofSocketRequest)}
+	b.write(r.Family)
+	b.write(r.Protocol)
+	b.write(r.Ext)
+	b.write(r.pad)
+	native.PutUint32(b.next(4), r.States)
+	networkOrder.PutUint16(b.next(2), r.ID.SourcePort)
+	networkOrder.PutUint16(b.next(2), r.ID.DestinationPort)
+	if r.Family == unix.AF_INET6 {
+		copy(b.next(16), r.ID.Source)
+		copy(b.next(16), r.ID.Destination)
+	} else {
+		copy(b.next(4), r.ID.Source.To4())
+		b.next(12)
+		copy(b.next(4), r.ID.Destination.To4())
+		b.next(12)
+	}
+	native.PutUint32(b.next(4), r.ID.Interface)
+	native.PutUint32(b.next(4), r.ID.Cookie[0])
+	native.PutUint32(b.next(4), r.ID.Cookie[1])
+	return b.Bytes
+}
+
+func (r *socketRequest) Len() int { return sizeofSocketRequest }
+
+type readBuffer struct {
+	Bytes []byte
+	pos   int
+}
+
+func (b *readBuffer) Read() byte {
+	c := b.Bytes[b.pos]
+	b.pos++
+	return c
+}
+
+func (b *readBuffer) Next(n int) []byte {
+	s := b.Bytes[b.pos : b.pos+n]
+	b.pos += n
+	return s
+}
+
+type socket netlink.Socket
+
+func (s *socket) deserialize(b []byte) error {
+	if len(b) < sizeofSocket {
+		return fmt.Errorf("socket data short read (%d); want %d", len(b), sizeofSocket)
+	}
+	rb := readBuffer{Bytes: b}
+	s.Family = rb.Read()
+	s.State = rb.Read()
+	s.Timer = rb.Read()
+	s.Retrans = rb.Read()
+	s.ID.SourcePort = networkOrder.Uint16(rb.Next(2))
+	s.ID.DestinationPort = networkOrder.Uint16(rb.Next(2))
+	if s.Family == unix.AF_INET6 {
+		s.ID.Source = net.IP(rb.Next(16))
+		s.ID.Destination = net.IP(rb.Next(16))
+	} else {
+		s.ID.Source = net.IPv4(rb.Read(), rb.Read(), rb.Read(), rb.Read())
+		rb.Next(12)
+		s.ID.Destination = net.IPv4(rb.Read(), rb.Read(), rb.Read(), rb.Read())
+		rb.Next(12)
+	}
+	s.ID.Interface = native.Uint32(rb.Next(4))
+	s.ID.Cookie[0] = native.Uint32(rb.Next(4))
+	s.ID.Cookie[1] = native.Uint32(rb.Next(4))
+	s.Expires = native.Uint32(rb.Next(4))
+	s.RQueue = native.Uint32(rb.Next(4))
+	s.WQueue = native.Uint32(rb.Next(4))
+	s.UID = native.Uint32(rb.Next(4))
+	s.INode = native.Uint32(rb.Next(4))
+	return nil
+}
+
+func destroySocket(sockId netlink.SocketID, family uint8, protocol uint8) error {
+	s, err := nl.Subscribe(unix.NETLINK_INET_DIAG)
+	if err != nil {
+		return err
+	}
+	defer s.Close()
+
+	req := nl.NewNetlinkRequest(SOCK_DESTROY, unix.NLM_F_REQUEST)
+	req.AddData(&socketRequest{
+		Family:   family,
+		Protocol: protocol,
+		States:   uint32(0xfff),
+		ID:       sockId,
+	})
+	err = s.Send(req)
+	if err != nil {
+		fmt.Printf("error in destroying socket: %v", sockId)
+	}
+	return err
+}
+
+func socketDiagUDPExecutor(family uint8, receiver func(message syscall.NetlinkMessage) error) error {
```

Comment:
As the `Caveats` section mentions, this implementation only lookups sockets in the host namespace. Do we encourage users to enable `bpf-lb-sock-hostns-only` if their workloads establish long-live TCP connection or connected UDP?

Reply:
That's actually a good suggestion. I'll extend the docs commit to mention about this flag.